### PR TITLE
aspects: Default languages=all when using aspects

### PR DESCRIPTION
**Pull request by pareksha**
This ensures that there is no need to pass `languages` argument
for bears supporting all languages by making a list of all languages
as default for bears when they are run via aspects or both
aspects and settings.

Closes https://github.com/coala/coala/issues/5655

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `corobo mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

File: coalib/bearlib/languages/__init__.py [hunk 1]
```diff
@@ -17,17 +17,22 @@
...
 from .definitions.PLSQL import PLSQL
+from .definitions.PowerShell import PowerShell
 from .definitions.Python import Python
 from .definitions.Ruby import Ruby
 from .definitions.Scala import Scala
 from .definitions.Swift import Swift
+from .definitions.Tcl import Tcl
 from .definitions.Vala import Vala
 from .definitions.TypeScript import TypeScript
+from .definitions.Unknown import Unknown
```
Comment (jayvdb) [inline coalib/bearlib/languages/__init__.py]: Move unknown to top. It is special.

And move these editions to a separate commit, so we can ack that bit and then work on the rest of the patch.

File: coalib/collecting/Collectors.py [hunk 2]
```diff
@@ -350,6 +351,17 @@ def get_all_bears_names():
     return [bear.name for bear in get_all_bears()]
 
 
+def get_all_languages():
```
Comment (jayvdb) [inline coalib/collecting/Collectors.py]: And arg to include Unknown, default False.

File: coalib/collecting/Collectors.py [hunk 3]
```diff
@@ -350,6 +351,17 @@ def get_all_bears_names():
...
+    """
+    Get a ``list`` of all languages supported by coala, including
+    ``Unknown`` language.
+    """
+    return Languages([
+        language_file[: -3] for language_file in
+        os.listdir('coalib/bearlib/languages/definitions/')
```
Comment (jayvdb) [inline coalib/collecting/Collectors.py]: They are all here, loaded, so dont look at the directory.

File: tests/collecting/CollectorsTest.py [hunk 4]
```diff
@@ -422,3 +424,38 @@ def test_get_all_bears_names(self):
...
+                 'Ruby',
+                 'Scala',
+                 'Shell',
+                 'Swift',
+                 'Tcl',
+                 'TypeScript',
+                 'Unknown',
+                 'Vala',
+                 'VisualBasic',
+                 'm4'})
```
Comment (jayvdb) [inline tests/collecting/CollectorsTest.py]: This list needs to be loaded from the directory to check all lang files were loaded

File: tests/bears/BearTest.py [hunk 5]
```diff
@@ -157,6 +157,13 @@ class aspectsFixOnlyTestBear(Bear, aspects={
     pass
 
 
+class aspectsGeneralTestBear(Bear, aspects={
```
Comment (jayvdb) [inline tests/bears/BearTest.py]: This should go in test_bears dir , after https://github.com/coala/coala/pull/5651

File: coalib/collecting/Collectors.py (hunk 3 above)
Comment (pareksha) [reply coalib/collecting/Collectors.py]: How are they loaded until I do something like `from coalib.bearlib.languages import Python` for every language?

File: tests/bears/BearTest.py (hunk 5 above)
Comment (pareksha) [reply tests/bears/BearTest.py]: I see, I'll keep it here for now else the tests won't pass. Would rebase as soon as you get the issue solved :+1:

File: tests/collecting/CollectorsTest.py (hunk 4 above)
Comment (pareksha) [reply tests/collecting/CollectorsTest.py]: If I load this from the directory, the versions won't come. It would be `Python` instead of `Python 2.7, 3.3, 3.4, 3.5, 3.6`. Then I have to use `Languages` for that, which would render the test useless.

One solution to check whether all files are loaded or not is to check that the number of language files in the directory are equal to the number of items supplied by `get_all_languages()`.

File: coalib/collecting/Collectors.py (hunk 3 above)
Comment (jayvdb) [reply coalib/collecting/Collectors.py]: See `coalib.bearlib.languages.__dict__` for one way to get them all

But manually listing them is also ok. The test will fail if someone forgets to add one.

File: tests/collecting/CollectorsTest.py (hunk 4 above)
Comment (jayvdb) [reply tests/collecting/CollectorsTest.py]: a count comparson sounds good.

File: coalib/bearlib/languages/__init__.py (hunk 1 above)
Comment (pareksha) [reply coalib/bearlib/languages/__init__.py]: Unknown was already imported, didn't see at first, removing this extra import

File: coalib/collecting/Collectors.py [hunk 6]
```diff
@@ -350,6 +352,18 @@ def get_all_bears_names():
     return [bear.name for bear in get_all_bears()]
 
 
+def get_all_languages(include_unknown=False):
+    """
+    Get a ``list`` of all languages supported by coala, including
+    ``Unknown`` language.
+    """
+    languages = [
+        key for key in definitions.__dict__ if not key.startswith('__')]
```
Comment (pareksha) [inline coalib/collecting/Collectors.py]: with `[key for key in coalib.bearlib.languages.definitions.__dict__]`, I was getting things like `__file__`, `__spec__` and some other dunders so I added the if statement. If there's another more efficient way to do it, please let me know. :)

File: coalib/collecting/Collectors.py [hunk 7]
```diff
@@ -350,6 +352,18 @@ def get_all_bears_names():
     return [bear.name for bear in get_all_bears()]
 
 
+def get_all_languages(include_unknown=False):
+    """
+    Get a ``list`` of all languages supported by coala, including
+    ``Unknown`` language.
```
Comment (jayvdb) [inline coalib/collecting/Collectors.py]: :Param:

File: coalib/collecting/Collectors.py (hunk 6 above)
Comment (jayvdb) [reply coalib/collecting/Collectors.py]: isinstance

File: tests/bears/BearTest.py (hunk 5 above)
Comment (jayvdb) [reply tests/bears/BearTest.py]: It is ready

Comment (pareksha) [reply tests/bears/BearTest.py]: Shifted `LANGUAGE_NAMES` and `LANGUAGE_COUNT` to `TestUtilities.py` as you did in  #5651.

File: coalib/collecting/Collectors.py (hunk 7 above)
Comment (pareksha) [reply coalib/collecting/Collectors.py]: Added :return: too :)

File: coalib/collecting/Collectors.py (hunk 6 above)
Comment (pareksha) [reply coalib/collecting/Collectors.py]: That helped a lot :+1:

File: tests/bears/BearTest.py (hunk 5 above)
Comment (jayvdb) [reply tests/bears/BearTest.py]: This Bear needs to be a separate module in our tests/test_bears/ directory

File: tests/coalaJSONTest.py [hunk 8]
```diff
@@ -187,5 +187,5 @@ def test_show_language_bears_output_file(self):
         os.remove('bears.json')
 
         self.assertEqual(retval, 0)
-        self.assertEqual(len(data['bears']), 2)
+        self.assertEqual(len(data['bears']), 3)
```
Comment (jayvdb) [inline tests/coalaJSONTest.py]: Create constant

Comment (pareksha) [reply tests/coalaJSONTest.py]: `JAVA_BEARS_CONSTANT = 3` as a module level constant?

File: tests/coalaTest.py [hunk 9]
```diff
@@ -186,8 +186,8 @@ def test_show_language_bears(self, debug=False):
             retval, stdout, stderr = execute_coala(
                 coala.main, 'coala', '-B', '-l', 'java', '-I', debug=debug)
             self.assertEqual(retval, 0)
-            # 2 bears plus 1 line holding the closing colour escape sequence.
-            self.assertEqual(len(stdout.splitlines()), 3)
+            # 3 bears plus 1 line holding the closing colour escape sequence.
+            self.assertEqual(len(stdout.splitlines()), 4)
```
Comment (jayvdb) [inline tests/coalaTest.py]: Create a constant , derived from the other constants

File: tests/parsing/FilterTest.py [hunk 10]
```diff
@@ -16,8 +16,8 @@ def test_filter_by_language_c(self):
             retval, stdout, stderr = execute_coala(
                 coala.main, 'coala', '-B', '--filter-by', 'language', 'c')
             self.assertEqual(retval, 0)
-            # 1 bear plus 1 line holding the closing colour escape sequence.
-            self.assertEqual(len(stdout.strip().splitlines()), 2)
+            # 2 bears plus 1 line holding the closing colour escape sequence.
+            self.assertEqual(len(stdout.strip().splitlines()), 3)
```
Comment (jayvdb) [inline tests/parsing/FilterTest.py]: As above

File: tests/settings/ConfigurationGatheringTest.py [hunk 11]
```diff
@@ -466,9 +466,9 @@ def test_get_filtered_bears(self):
             local_bears, global_bears = get_filtered_bears(
```
Comment (jayvdb) [inline tests/settings/ConfigurationGatheringTest.py]: @MacBox7 this is unsorted?

We need to apply a sort here I guess otherwise these will jiggle around randomly

Comment (MacBox7) [reply tests/settings/ConfigurationGatheringTest.py]: Either we can sort the output of `get_filtered_bears` or we can provide a utility in `TestUtilities.py`. Which one do you think would be best?

Comment (pareksha) [reply tests/settings/ConfigurationGatheringTest.py]: Will it be a part of this PR?

File: tests/settings/ConfigurationGatheringTest.py [hunk 12]
```diff
@@ -470,9 +470,9 @@ def test_get_filtered_bears(self):
         local_bears['cli'] = _sort_bears(local_bears['cli'])
         global_bears['cli'] = _sort_bears(global_bears['cli'])
 
-        self.assertEqual(len(local_bears['cli']), 2)
-        self.assertEqual(str(local_bears['cli'][0]),
-                         "<class 'JavaTestBear.JavaTestBear'>")
+        self.assertEqual(len(local_bears['cli']), 3)
         self.assertEqual(str(local_bears['cli'][1]),
```
Comment (MacBox7) [inline tests/settings/ConfigurationGatheringTest.py]: Why not assert the instances of all three bears.

Comment (pareksha) [reply tests/settings/ConfigurationGatheringTest.py]: Good idea, made the change :+1:

**Comment by jayvdb** [comment]
ack https://github.com/coala/coala/pull/5656/commits/e83dfcc1f0f00cf7aa9abf4f51ef7a6effa804d4 https://github.com/coala/coala/pull/5656/commits/656ed289b5a4ab2660e848f409c53256f59dda0a https://github.com/coala/coala/pull/5656/commits/0986d56140a7a5e57c97928e9477cc4ac74e73b8
